subsys/bluetooth/host/att.c

```diff
@@ -189,11 +189,18 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf,
 
 	chan->sent = cb ? cb : chan_cb(buf);
 
+	/* bt_l2cap_send_cb takes onwership of the buffer so take another
+	 * reference to restore the state in case an error is returned.
+	 */
+	net_buf_ref(buf);
+
 	err = bt_l2cap_send_cb(chan->att->conn, BT_L2CAP_CID_ATT,
 			       buf, att_cb(chan->sent),
 			       &chan->chan.chan);
 	if (err) {
 		net_buf_simple_restore(&buf->b, &state);
+	} else {
+		net_buf_unref(buf);
```

Comment:
This all makes for very unintuitive and confusing looking code (which is true any time you have conditional unrefs), even with the code comment. Would it make sense to consider consider changing the semantics of `bt_l2cap_send_cb()` to either:

 - Always grabs a new reference to itself (i.e. caller is always responsible for unrefing)

or

 - Only take ownership in case of success. In fact, looking at the callers of `chan_send()` it looks like that's the assumption for chan send, i.e. the callers assume they still own buff when they get an error.

Reply:
I think second option makes more sense. 

@Vudentz Does the same problem exist in `l2cap_chan_le_send`?
There is a `bt_l2cap_send_cb` and a `net_buf_simple_restore` there as well.

The ref handling easily leads to these kinds of mistakes, it's not the first time.

Reply:
> I think second option makes more sense.
> 
> @Vudentz Does the same problem exist in `l2cap_chan_le_send`?
> There is a `bt_l2cap_send_cb` and a `net_buf_simple_restore` there as well.
> 
> The ref handling easily leads to these kinds of mistakes, it's not the first time.

Yeah, if seg == buf it looks like there would be indeed a similar problem.

Reply:
We could perhaps introduce a version that does not unref on error, e.g. bt_conn_send_nounref so callers that want to handler the errors themselves can use that instead of the regular one that always unref, thougths?

Reply:
@jhedberg @joerchan I took the approach of changing the default behavior to not unref in case of failure since there are more code that rely on that and then introduced bt_conn_send_unref(_cb) for the code that want to stick to the old behavior.